app/src/main/scala/org/alephium/explorer/service/sync/SimersSyncPlayground.scala

```diff
@@ -0,0 +1,312 @@
+// along with the library. If not, see <http://www.gnu.org/licenses/>.
+
+package org.alephium.explorer.service.sync
+
+
+import com.typesafe.scalalogging.StrictLogging
+import org.alephium.explorer.{foldFutures, GroupSetting}
+import org.alephium.explorer.api.model.{GroupIndex, Height}
+import org.alephium.explorer.cache.BlockCache
+import org.alephium.explorer.persistence.queries.BlockQueries
+import org.alephium.explorer.persistence.DBRunner._
+import org.alephium.explorer.persistence.dao.BlockDao
+import org.alephium.explorer.persistence.model.BlockEntity
+import org.alephium.explorer.service.BlockFlowClient
+import slick.basic.DatabaseConfig
+import slick.jdbc.PostgresProfile
+
+import scala.collection.immutable.ArraySeq
+import scala.concurrent.{ExecutionContext, Future}
+
+/**
+  * My Sync test-bed.
+  */
+object SimersSyncPlayground extends StrictLogging {
+
+  /** Another version of [[init]]. Returns a boolean, ignores any [[Future]] failures */
+  def initIgnoreError()(implicit ec: ExecutionContext,
+                        dc: DatabaseConfig[PostgresProfile],
+                        client: BlockFlowClient,
+                        groupSetting: GroupSetting,
+                        cache: BlockCache): Future[Boolean] =
+    init()
+      .map(_ => true)
+      .recover {
+        error =>
+          logger.error("Failed to initialise sync", error)
+          false
+      }
+
+  /** The init function from `BlockFlowSyncService` */
+  def init()(implicit ec: ExecutionContext,
+             dc: DatabaseConfig[PostgresProfile],
+             client: BlockFlowClient,
+             groupSetting: GroupSetting,
+             cache: BlockCache): Future[Unit] =
+    for {
+      heightAndGroups <- run(BlockQueries.maxHeightZipped(groupSetting.groupIndexes))
+      bestBlocks <- client.fetchRootBlockAtHeights(remoteRootBlocksToFetch(heightAndGroups))
+      result <- insert(bestBlocks)
+    } yield result
+
+  /**
+    * Give a set of blocks, find which blocks are out of sync and then persist
+    * them in a single transaction.
+    *
+    * TODO: return the number of blocks synced
+    * */
+  def insert(blocks: ArraySeq[BlockEntity])(implicit ec: ExecutionContext,
+                                            dc: DatabaseConfig[PostgresProfile],
+                                            client: BlockFlowClient,
+                                            groupSetting: GroupSetting,
+                                            cache: BlockCache): Future[Unit] =
+    if (blocks.isEmpty)
+      Future.unit
+    else
+      for {
+        outOfSyncBlocks <- findOutOfSyncBlocks(blocks) //find out-of-sync blocks
+        missingBlocks <- executeActions(outOfSyncBlocks) //persist out of sync blocks
+        fetchedMissingBlocks <- client.fetchMissingBlocks(missingBlocks) //updateMainChain returns missing blocks
+        result <- insert(fetchedMissingBlocks) //persist those missing blocks
```

Comment:
I can't find the place where you actually insert the `blocks`

Reply:
This is actually an issue for the `findOutOfSyncBlocks` functions.
Let say we have 
```
B0 -> B1 
```
already in DB and we need to insert 
```
B2->B3->B4
```
when calling `findOutOfSyncBlocks` on the 3 new blocks, for `B2` it's fine as it's parent `B1` is alreay in db, even if it was not, it's correct to fetch it. But as we don't (at least I can't see it) insert `B2`, when calling `findOutOfSyncBlock` on `B3`, it won't find it's parent `B2` as it wasn't inserted, we will thus refetch it for nothing.
Correct me if I miss something.

Reply:
> I can't find the place where you actually insert the blocks
The insert happens in one place [here](https://github.com/alephium/explorer-backend/pull/362/files#diff-6e6921810a0484ff6cfb7cb484079a5f39b570b61ac07573f4ad79549134dcedR167).

> B2->B3->B4

You are totally right. This is not something I've gotten around to yet, to avoid double fetching, but it has a simple solution. 

This PR was like my last attempt at understanding the sync process. Was having trouble understanding whats going on under the hood. Still unsure if I fully understand it. 

Totally agree with you. Chat on Slack.